### PR TITLE
[diffusion] feat: enable compile warmup for vae decode

### PR DESCRIPTION
**Pull request by mickqian**
## Summary

This PR makes `--enable-torch-compile` cover the diffusion decode hot path without making the first live request pay compile latency.

Changes:
- compile VAE `decode` from `DecodingStage` when `--enable-torch-compile` is enabled
- use a VAE-specific default compile mode of `default`, with overrides via `SGLANG_VAE_TORCH_COMPILE_MODE` or `SGLANG_TORCH_COMPILE_MODE`
- automatically enable server warmup for torch compile when no warmup mode is explicitly configured
- preserve explicit `--warmup-steps` for torch-compile server warmup
- add internal non-warmup pretraffic requests before the server is marked ready, so the real request path is also compiled
- fix Z-Image caption conditioning so DiT receives bucket-padded caption tensors plus tensor valid lengths, avoiding per-prompt-length graph recompiles

## Why

VAE compile improves steady-state decode latency, but the first compiled VAE call can be very expensive. Separately, a normal warmup request was not enough for first live traffic because warmup and real requests can exercise different branches. Z-Image also exposed a prompt-length guard issue: caption embeddings were padded inside the compiled transformer forward, so different raw prompt lengths triggered new graphs even when they belonged to the same padded bucket.

## Validation

Remote validation on a non-CI 1x H100 rx devbox, base `origin/main` `bc150173b`.

Unit tests:

```bash
python3 -m pytest \
  python/sglang/multimodal_gen/test/unit/test_server_args.py::TestWarmupModeNormalization \
  python/sglang/multimodal_gen/test/unit/test_cfg_parallel_warmup.py::TestWarmupReqCfgParallel \
  -q
```

Result: `37 passed`.

Pre-commit:

```bash
pre-commit run --files \
  python/sglang/multimodal_gen/configs/pipeline_configs/zimage.py \
  python/sglang/multimodal_gen/runtime/managers/scheduler.py \
  python/sglang/multimodal_gen/runtime/models/dits/zimage.py \
  python/sglang/multimodal_gen/runtime/pipelines_core/stages/decoding.py \
  python/sglang/multimodal_gen/runtime/server_args.py \
  python/sglang/multimodal_gen/runtime/server_warmup.py \
  python/sglang/multimodal_gen/runtime/warmup_request_builder.py \
  python/sglang/multimodal_gen/test/unit/test_cfg_parallel_warmup.py \
  python/sglang/multimodal_gen/test/unit/test_server_args.py
```

Result: passed.

Server smoke:

```bash
sglang serve \
  --model-path Tongyi-MAI/Z-Image-Turbo \
  --host 127.0.0.1 \
  --port 31000 \
  --attention-backend torch_sdpa \
  --enable-torch-compile \
  --warmup-resolutions 1024x1024 \
  --warmup-steps 1
```

Request: `1024x1024`, `num_inference_steps=1`, `guidance_scale=0.0`.

Observed cold warmup and first external request:
- cold server warmup request: `123.75s`
- internal real-path prewarm denoise: `1.0928s`, then `0.0220s`
- first external real request after ready: wall `0.383s`, `DenoisingStage 0.0315s`, `DecodingStage 0.0098s`

Note: `torch_sdpa` was used only because this H100 validation environment has a local FA3 binding mismatch in the text encoder path; this PR does not depend on SDPA.

---
### CI States

Latest PR Test (Base): :white_check_mark: [Run #28694204030](https://github.com/sgl-project/sglang/actions/runs/28694204030)
Latest PR Test (Extra): :x: [Run #28694203997](https://github.com/sgl-project/sglang/actions/runs/28694203997)

### CODE REVIEW

**Comment by mickqian** [comment]
Baseline vs PR data after rebasing on latest `origin/main`.

Setup:
- Machine: clean H100 devbox, 1 GPU
- Model: `Tongyi-MAI/Z-Image-Turbo`
- Request: `1024x1024`, `num_inference_steps=1`, `guidance_scale=0`, `seed=46`
- Server args: `--enable-torch-compile --warmup-resolutions 1024x1024 --warmup-steps 1 --attention-backend torch_sdpa`
- Torch compile cache cleared per run

Refs:
- Baseline: `origin/main` `430418e2182606b5296b26c912df8f12ec9b380f`
- PR: `a6067f52356195324bdbd52c1769485639eb7e02`

Client latency, sending immediately after `/health`:

| variant | request 1 | request 2 | request 3 | note |
|---|---:|---:|---:|---|
| baseline | 92.766s | 0.358s | 0.355s | request 1 waits for warmup and then pays first real-path compile |
| PR | 110.616s | 0.288s | 0.287s | request 1 waits for longer compile warmup + internal prewarm |

Client latency, waiting 150s after `/health` before first external request:

| variant | request 1 | request 2 | request 3 |
|---|---:|---:|---:|
| baseline | 4.493s | 0.361s | 0.359s |
| PR | 0.367s | 0.313s | 0.312s |

Server-side stage evidence for the post-warmup run:

| variant | first external denoise | first external decode |
|---|---:|---:|
| baseline | 4.1125s | 0.0091s |
| PR | 0.0342s | 0.0099s |

Interpretation: the PR does not reduce cold compile cost; it moves it into server warmup and adds real-path prewarm so the first external request after warmup is close to steady state. The immediate-after-`/health` number is still high because `/health` returns before background warmup finishes, and the request waits behind warmup.

**Comment by mickqian** [comment]
Corrected benchmark focus: steady-state VAE decode time, not first external request latency.

Setup:
- H100 devbox, 1 GPU
- `Tongyi-MAI/Z-Image-Turbo`, `1024x1024`, `num_inference_steps=1`, `guidance_scale=0`
- server args include `--enable-torch-compile --warmup-resolutions 1024x1024 --warmup-steps 1 --attention-backend torch_sdpa`
- each run clears the torch compile cache, waits for `/health`, waits another 150s, then sends 12 external requests
- primary steady-state metric below excludes request 1 and reports request 2-12

| variant | commit | steady DecodingStage mean | median | p90 | steady client E2E mean | median |
|---|---:|---:|---:|---:|---:|---:|
| baseline `origin/main` | `430418e` | 6.96 ms | 6.30 ms | 9.30 ms | 367.05 ms | 367.23 ms |
| PR | `a6067f52` | 5.21 ms | 5.10 ms | 6.80 ms | 297.99 ms | 297.40 ms |

Steady decode speedup: 1.34x by mean, 1.24x by median. Steady client E2E speedup: 1.23x by mean.

Request 1 is intentionally not the headline number here: in this run baseline still had a first external denoise compile outlier after the wait, while the PR prewarm path did not.

**Comment by mickqian** [comment]
Additional cold-cache compile/warmup cost profile.

Setup:
- same H100 devbox / Z-Image-Turbo 1024x1024 setup
- each run starts from an idle GPU and clears `SGLANG_DIFFUSION_CACHE_ROOT/torch_compile_cache`
- no reinstall in the measured window
- one external request is sent immediately after `/health`; it waits behind server warmup when warmup is still running
- peak VRAM is sampled by `nvidia-smi` at 0.2s intervals

| variant | commit | server warmup compile window | `/health` time | first request latency while waiting for warmup | nvidia-smi peak VRAM | request-reported peak |
|---|---:|---:|---:|---:|---:|---:|
| baseline `origin/main` | `430418e` | 71.96s | 38.20s | 75.10s | 24,349 MiB | 17,842 MiB |
| PR | `a6067f52` | 81.45s | 39.20s | 83.88s | 21,877 MiB | 15,370 MiB |

The PR adds about 9.5s to the cold server warmup compile window in this run, because it also precompiles/prewarms the VAE decode path. The clean profile does not show a peak-VRAM regression.

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/decoding.py [hunk 1]
```diff
@@ -155,6 +159,33 @@ def verify_output(self, batch: Req, server_args: ServerArgs) -> VerificationResu
...
         return scale_and_shift_latents(latents, server_args, self.vae)
 
+    def _get_vae_decode_fn(self, vae, server_args: ServerArgs):
+        if not server_args.enable_torch_compile or not isinstance(vae, nn.Module):
+            return vae.decode
+
+        vae_id = id(vae)
+        decode_fn = self._torch_compiled_vae_decode_by_id.get(vae_id)
+        if decode_fn is not None:
+            return decode_fn
```
[the remark this answers is not in the document]
Comment (mickqian) [reply python/sglang/multimodal_gen/runtime/pipelines_core/stages/decoding.py]: Fixed in `c5d9ce2561`: the compiled VAE decode cache now clears when the active VAE instance changes, so the cached compiled bound method does not keep older VAE instances alive. Added `test_torch_compile_decode_cache_is_replaced_for_new_vae_instance`.

File: python/sglang/multimodal_gen/runtime/managers/scheduler.py [hunk 2]
```diff
@@ -1048,7 +1048,12 @@ def event_loop(self) -> None:
                     is_warmup = is_warmup_req(processed_req)
                     self._log_warmup_result(output_batch, processed_req, is_warmup)
 
-                    if is_warmup and should_return_warmup_result(processed_req):
+                    if processed_req.extra.get("server_internal_prewarm"):
```
[the remark this answers is not in the document]
Comment (mickqian) [reply python/sglang/multimodal_gen/runtime/managers/scheduler.py]: Fixed in `d4449cd263`: scheduler now uses `get_first_generation_req()` before reading request metadata, so LoRA/control requests without `extra` are ignored by this warmup-result path. Added regression coverage for `SetLoraReq` and `UnmergeLoraWeightsReq`.

File: python/sglang/multimodal_gen/runtime/warmup_request_builder.py [hunk 3]
```diff
@@ -340,12 +353,31 @@ def build_warmup_reqs(
...
+        )
+        prompts = (
+            (DEFAULT_PLACEHOLDER_PROMPT,) + TORCH_COMPILE_REAL_PATH_PREWARM_PROMPTS
+            if run_real_path_prewarm
+            else (DEFAULT_PLACEHOLDER_PROMPT,)
+        )
+        for prompt_idx, prompt in enumerate(prompts):
+            prompt_req_kwargs = req_kwargs.copy()
+            prompt_req_kwargs["prompt"] = prompt
+            req = Req(**prompt_req_kwargs)
```
[the remark this answers is not in the document]
Comment (mickqian) [reply python/sglang/multimodal_gen/runtime/warmup_request_builder.py:373]: Fixed in `c5d9ce2561`: each generated warmup request now gets its own copied `SamplingParams`, and the unit test verifies that mutating the internal prewarm request does not affect the formal warmup request.
